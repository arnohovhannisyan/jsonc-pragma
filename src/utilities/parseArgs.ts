import { IArguments } from "../models/arguments";

export function parseArgs(match: string | null): IArguments {
	if (!match) return {};

	const entries = Array.from(
		new URLSearchParams(match.replace(/\s(\w*?=)/, "&$1"))
	);

	return entries.reduce(
		(acc, [key, value]) => ({
			...acc,
			[key]: value
		}),
		{}
	);
}
