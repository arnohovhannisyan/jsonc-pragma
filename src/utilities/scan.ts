import matchBracket from "match-bracket";
import { ISection } from "../models/section";
import { parseArgs } from "./parseArgs";

function isRegular(line: string) {
	if (/[{[]$/.test(line.trim())) return false;

	return true;
}

/**
 * @param contents The contents of your JSON file
 */
export function scan(contents: string): ISection[] {
	const pragmaRegex = /\s*\/\/\s@(?<name>[a-zA-Z\d]+)(\s(?<args>.*))?$/;

	const lines = contents.split("\n");

	const linesWithPragmas = lines.filter(line => pragmaRegex.test(line));

	return linesWithPragmas.map<ISection>(line => {
		const startingLineNumber = lines.indexOf(line) + 1;
		const startingLine = lines[startingLineNumber];

		const groups = pragmaRegex.exec(line)!.groups!;

		if (isRegular(startingLine)) {
			return {
				start: startingLineNumber,
				end: startingLineNumber,
				args: parseArgs(groups.args),
				name: groups.name
			};
		}

		const startingBracketPos = {
			line: startingLineNumber + 1,
			cursor: startingLine.trimEnd().length
		};

		const endingLineNumber =
			matchBracket(contents.toString(), startingBracketPos).line - 1;

		return {
			start: startingLineNumber,
			end: endingLineNumber,
			args: parseArgs(groups.args),
			name: groups.name
		};
	});
}
