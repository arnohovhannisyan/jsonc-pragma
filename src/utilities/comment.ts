import { ISection } from "~/models";
import { scan } from "~/utilities/scan";

/**
 * @param contents The contents of your JSON file
 * @param selector A function that accepts a section and determines if it should be commented by returning true (comment) or false (don't comment).
 */
export function comment(
	contents: string | Buffer,
	selector: (section: ISection) => boolean = () => true
): string {
	const sections = scan(contents);

	const sectionsToComment = sections.filter(selector);

	const lines = contents.toString().split("\n");

	return lines
		.map((line, i) => {
			if (line.trimStart().startsWith("//")) return line;

			const section = sectionsToComment.find(s => i >= s.start && i <= s.end);

			if (section) {
				const firstLine = lines[section.start];
				const leadingSpaces = /^(\s*)/.exec(firstLine)?.[0].length ?? 0;

				return line.replace(new RegExp(`^(\\s{${leadingSpaces}})`), "$1// ");
			}

			return line;
		})
		.join("\n");
}
