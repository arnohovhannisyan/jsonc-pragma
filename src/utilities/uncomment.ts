import { ISection } from "../models/section";
import { scan } from "./scan";

/**
 * @param contents The contents of your JSON file
 * @param selector A function that accepts a section and determines if it should be uncommented by returning true (uncomment) or false (don't uncomment).
 */
export function uncomment(
	contents: string,
	selector: (section: ISection) => boolean = () => true
): string {
	const sections = scan(contents.toString().replace(/\/\/\s(?!@)/g, ""));

	const sectionsToUncomment = sections.filter(selector);

	const uncommented = contents
		.split("\n")
		.map((line, i) => {
			const shouldUncomment = sectionsToUncomment.some(
				section => i >= section.start && i <= section.end
			);

			if (shouldUncomment) return line.replace("// ", "");

			return line;
		})
		.join("\n");

	return uncommented;
}
