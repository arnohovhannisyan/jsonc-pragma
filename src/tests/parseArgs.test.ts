import { parseArgs } from "../utilities/parseArgs";

test("basic functionality", () => {
	expect(parseArgs("abc=5")).toMatchSnapshot();
	expect(parseArgs("abc=hello")).toMatchSnapshot();
	expect(parseArgs(null)).toMatchSnapshot();
});
