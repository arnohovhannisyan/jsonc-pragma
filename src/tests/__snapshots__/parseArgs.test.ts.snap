// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic functionality 1`] = `
Object {
  "abc": "5",
}
`;

exports[`basic functionality 2`] = `
Object {
  "abc": "hello",
}
`;

exports[`basic functionality 3`] = `Object {}`;
